openmdao/core/component.py

```diff
@@ -631,3 +635,116 @@ def _get_relname_map(self, parent_unknowns):
+            are taken.
+
+        fd_unknowns : list of strings, optional
+            List of output or state name strings for derivatives to be
+            calculated. This is used by problem to limit the derivatives that
+            are taken.
+
+        poi_indices: dict of list of integers, optional
+            Should be an empty list, as there is no subcomponent relevance reduction.
+
+        qoi_indices: dict of list of integers, optional
+            Should be an empty list, as there is no subcomponent relevance reduction.
+
+        Returns
+        -------
+        dict
+            Dictionary whose keys are tuples of the form ('unknown', 'param')
+            and whose values are ndarrays containing the derivative for that
+            tuple pair.
+        """
+
+        # Params and Unknowns that we provide at this level.
+        if fd_params is None:
+            fd_params = self._get_fd_params()
+        if fd_unknowns is None:
+            fd_unknowns = self._get_fd_unknowns()
+
+        # Use settings in the system dict unless variables override.
+        step_size = self.fd_options.get('step_size', 1.0e-6)
+
+        jac = {}
+        csparams = ComplexStepTgtVecWrapper(params)
+        csunknowns = ComplexStepSrcVecWrapper(unknowns)
+        csresids = ComplexStepSrcVecWrapper(resids)
+
+        # Pull result from resids only if comp overrides apply_nonlinear
+        states = self.states
+        if len(states) > 0:
+            resultvec = csresids
+        else:
+            resultvec = csunknowns
+
+        # Compute gradient for this param or state.
+        for p_name in chain(fd_params, states):
+
+            # States are stepped in unknowns, not params
+            if p_name in states:
+                stepvec = csunknowns
+                target_input = unknowns.flat[p_name]
+            else:
+                stepvec = csparams
+                target_input = params.flat[p_name]
+
+            stepvec.set_complex_var(p_name)
+
+            mydict = {}
+            if p_name in self._to_abs_pnames:
+                for val in itervalues(self._params_dict):
+                    if val['promoted_name'] == p_name:
+                        mydict = val
+                        break
+
+            # Local settings for this var trump all
+            fdstep = mydict.get('step_size', step_size)
+
+            # Size our Inputs
+            p_size = np.size(target_input)
+            p_idxs = range(p_size)
+
+            # Size our Outputs
+            for u_name in fd_unknowns:
+                u_size = np.size(unknowns[u_name])
+                jac[u_name, p_name] = np.zeros((u_size, p_size))
+
+            # apply Complex Step on each index in array
+            for j, idx in enumerate(p_idxs):
+
+                stepvec.step_complex(idx, fdstep)
+                self.apply_nonlinear(csparams, csunknowns, csresids)
+                stepvec.step_var = None
```

Comment:
Pushed up a fix for this. Added a function to unset.  We need to unset because we may swap to different vector on the next input. States and Params need to step in different vectors. I did move the location of the statement, since it was not in the proper location.